pkg/storegateway/bucket.go

```diff
@@ -965,71 +953,64 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storepb.Store_Serie
 
 	s.mtx.RLock()
 
-	for _, bs := range s.blockSets {
-		blockMatchers, ok := bs.labelMatchers(matchers...)
-		if !ok {
-			continue
-		}
+	blocks := s.blockSet.getFor(req.MinTime, req.MaxTime, req.MaxResolutionWindow, reqBlockMatchers)
 
-		blocks := bs.getFor(req.MinTime, req.MaxTime, req.MaxResolutionWindow, reqBlockMatchers)
+	if s.debugLogging {
+		debugFoundBlockSetOverview(s.logger, req.MinTime, req.MaxTime, req.MaxResolutionWindow, blocks)
+	}
 
-		if s.debugLogging {
-			debugFoundBlockSetOverview(s.logger, req.MinTime, req.MaxTime, req.MaxResolutionWindow, bs.labels, blocks)
-		}
+	for _, b := range blocks {
+		b := b
 
-		for _, b := range blocks {
-			b := b
+		if s.enableSeriesResponseHints {
+			// Keep track of queried blocks.
+			resHints.AddQueriedBlock(b.meta.ULID)
+		}
 
-			if s.enableSeriesResponseHints {
-				// Keep track of queried blocks.
-				resHints.AddQueriedBlock(b.meta.ULID)
-			}
+		var chunkr *bucketChunkReader
+		// We must keep the readers open until all their data has been sent.
+		indexr := b.indexReader()
+		if !req.SkipChunks {
+			chunkr = b.chunkReader(gctx)
+			defer runutil.CloseWithLogOnErr(s.logger, chunkr, "series block")
+		}
 
-			var chunkr *bucketChunkReader
-			// We must keep the readers open until all their data has been sent.
-			indexr := b.indexReader()
-			if !req.SkipChunks {
-				chunkr = b.chunkReader(gctx)
-				defer runutil.CloseWithLogOnErr(s.logger, chunkr, "series block")
-			}
+		// Defer all closes to the end of Series method.
+		defer runutil.CloseWithLogOnErr(s.logger, indexr, "series block")
 
-			// Defer all closes to the end of Series method.
-			defer runutil.CloseWithLogOnErr(s.logger, indexr, "series block")
+		// If query sharding is enabled we have to get the block-specific series hash cache
+		// which is used by blockSeries().
+		var blockSeriesHashCache *hashcache.BlockSeriesHashCache
+		if shardSelector != nil {
+			blockSeriesHashCache = s.seriesHashCache.GetBlockCache(b.meta.ULID.String())
+		}
 
-			// If query sharding is enabled we have to get the block-specific series hash cache
-			// which is used by blockSeries().
-			var blockSeriesHashCache *hashcache.BlockSeriesHashCache
-			if shardSelector != nil {
-				blockSeriesHashCache = s.seriesHashCache.GetBlockCache(b.meta.ULID.String())
+		g.Go(func() error {
+			part, pstats, err := blockSeries(
+				gctx,
+				indexr,
+				chunkr,
+				matchers,
```

Comment:
_Note to reviewers_

In Mimir use case, `matchers` is always equal to `blockMatchers`. See my note on `labelMatchers()` for more details.